File: stacks/nodejs/react-fastify/rest/bff/src/flow/inference/waitingOnEvidence.ts
import { degradeConfidence, resolveActors, resolvePrimaryTeam } from "./explain";
import type { ProviderAdapterInput, NormalizedEvidenceState } from "../types";
import type { FlowSignal } from "./types";

function selectEvidence(evidenceStates: NormalizedEvidenceState[]): NormalizedEvidenceState | null {
  const candidates = evidenceStates.filter((evidence) =>
    evidence.state === "required" || evidence.state === "pending" || evidence.state === "stale"
  );
  if (candidates.length === 0) {
    return null;
  }
  return candidates.sort((a, b) => Date.parse(b.as_of) - Date.parse(a.as_of))[0];
}

export function inferWaitingOnEvidence(input: ProviderAdapterInput): FlowSignal | null {
  if (!input.evidence_states || input.evidence_states.length === 0) {
    return null;
  }

  const evidence = selectEvidence(input.evidence_states);
  if (!evidence) {
    return null;
  }

  const change = input.changes.find((c) => c.provider_id === evidence.change_id);
  if (!change || change.state !== "merged") {
    return null;
  }

  const runsForChange = input.validation_runs.filter((run) => run.change_id === change.provider_id);
  const hasPassedValidation = runsForChange.some((run) => run.state === "passed");
  const hasBlockingValidation = runsForChange.some((run) =>
    run.state === "failed" || run.state === "flaky" || run.state === "pending" || run.state === "running",
  );

  if (!hasPassedValidation || hasBlockingValidation) {
    return null;
  }

  const ownerNames = resolveActors(input.actors, evidence.owner_actor_id ? [evidence.owner_actor_id] : []);
  const evidenceTypes = (evidence.required_types ?? []).join(", ");
  const service = input.repository.full_name ?? input.repository.provider_id;
  const ownerActor = input.actors.find((actor) => actor.provider_id === evidence.owner_actor_id);
  const team = ownerActor?.team_memberships?.[0] ?? resolvePrimaryTeam(input.ownership_hints);

  const partialFlag =
    evidence.is_partial === true ||
    change.is_partial === true ||
    input.repository.is_partial === true ||
    runsForChange.some((run) => run.is_partial);

  return {
    id: "waiting_on_evidence",
    title: "Waiting on evidence, not effort",
    severity: "medium",
    confidence: degradeConfidence("high", partialFlag),
    explanation:
      evidenceTypes.length > 0
        ? `Evidence pending for ${service}: ${evidenceTypes}.`
        : `Evidence pending for ${service}.`,
    recommendedNextAction:
      ownerNames.length > 0
        ? `Request required evidence from ${ownerNames.join(", ")}.`
        : "Request required evidence from the accountable owner.",
    relatedEntities: [service, change.provider_id, ...(evidence.required_types ?? [])],
    scope: { service, team, stage: "evidence" },
    observedAt: evidence.as_of ?? change.updated_at ?? input.repository.fetched_at,
  };
}

File: stacks/go/net-http/rest/bff/flow/inference/waiting_on_evidence.go
package inference

import (
	"fmt"
	"idp-go-net-http-rest/bff/flow"
	"sort"
	"strings"
)

func selectEvidence(states []flow.NormalizedEvidenceState) (flow.NormalizedEvidenceState, bool) {
	candidates := make([]flow.NormalizedEvidenceState, 0)
	for _, state := range states {
		if state.State == "required" || state.State == "pending" || state.State == "stale" {
			candidates = append(candidates, state)
		}
	}
	if len(candidates) == 0 {
		return flow.NormalizedEvidenceState{}, false
	}
	sort.Slice(candidates, func(i, j int) bool {
		return candidates[i].AsOf > candidates[j].AsOf
	})
	return candidates[0], true
}

func resolveActorName(actors []flow.NormalizedActor, id string) string {
	for _, actor := range actors {
		if actor.ProviderID == id {
			return actor.DisplayName
		}
	}
	return ""
}

func InferWaitingOnEvidence(input flow.ProviderAdapterInput) (*flow.FlowSignal, bool) {
	if len(input.EvidenceStates) == 0 {
		return nil, false
	}

	evidence, ok := selectEvidence(input.EvidenceStates)
	if !ok {
		return nil, false
	}

	var change *flow.NormalizedChange
	for i := range input.Changes {
		if input.Changes[i].ProviderID == evidence.ChangeID {
			change = &input.Changes[i]
			break
		}
	}
	if change == nil || change.State != "merged" {
		return nil, false
	}

	runs := make([]flow.NormalizedValidationRun, 0)
	for _, run := range input.ValidationRuns {
		if run.ChangeID == change.ProviderID {
			runs = append(runs, run)
		}
	}

	hasPassed := false
	hasBlocking := false
	for _, run := range runs {
		if run.State == "passed" {
			hasPassed = true
		}
		if run.State == "failed" || run.State == "flaky" || run.State == "pending" || run.State == "running" {
			hasBlocking = true
		}
	}
	if !hasPassed || hasBlocking {
		return nil, false
	}

	owner := resolveActorName(input.Actors, evidence.OwnerActorID)
	types := evidence.RequiredTypes
	service := serviceFrom(input.Repository)
	desc := fmt.Sprintf("Evidence pending for %s.", service)
	if len(types) > 0 {
		desc = fmt.Sprintf("Evidence pending for %s: %s.", service, strings.Join(types, ", "))
	}

	partial := evidence.IsPartial || change.IsPartial || input.Repository.IsPartial
	for _, run := range runs {
		if run.IsPartial {
			partial = true
			break
		}
	}
	team := resolvePrimaryTeam(input.OwnershipHints)
	if team == "" {
		for _, actor := range input.Actors {
			if actor.ProviderID == evidence.OwnerActorID && len(actor.TeamMemberships) > 0 {
				team = actor.TeamMemberships[0]
				break
			}
		}
	}

	signal := flow.FlowSignal{
		ID:          "waiting_on_evidence",
		Title:       "Waiting on evidence, not effort",
		Severity:    "medium",
		Confidence:  degradeConfidence("high", partial),
		Explanation: desc,
		RecommendedNextAction: func() string {
			if owner != "" {
				return fmt.Sprintf("Request required evidence from %s.", owner)
			}
			return "Request required evidence from the accountable owner."
		}(),
		RelatedEntities: []any{service, change.ProviderID},
		Scope: &flow.FlowSignalScope{
			Service: service,
			Team:    team,
			Stage:   flow.StageEvidence,
		},
		ObservedAt: evidence.AsOf,
	}
	for _, t := range types {
		signal.RelatedEntities = append(signal.RelatedEntities, t)
	}

	return &signal, true
}
